File: ui/src/main/java/com/example/quarter/ui/MainActivity.java
package com.example.quarter.ui;

import android.app.ProgressDialog;
import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.RecyclerView;
import com.example.quarter.domain.Note;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.snackbar.Snackbar;
import dagger.android.support.DaggerAppCompatActivity;

import javax.inject.Inject;
import java.util.List;

public class MainActivity extends DaggerAppCompatActivity implements MainView {

    @Inject
    MainPresenter presenter;

    @Inject
    NotesAdapter adapter;

    private ProgressDialog progressDialog;
    private ViewGroup rootLayout;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        setSupportActionBar((Toolbar) findViewById(R.id.toolbar));
        FloatingActionButton fab = (FloatingActionButton) findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                presenter.addNote();
            }
        });
        RecyclerView recyclerView = (RecyclerView) findViewById(R.id.recyclerView);
        recyclerView.setAdapter(adapter);
        presenter.setView(this);
        presenter.loadNotes();
        rootLayout = (ViewGroup) findViewById(R.id.rootLayout);
    }

    @Override
    public void showLoading() {
        progressDialog = new ProgressDialog(this);
        progressDialog.setProgressStyle(ProgressDialog.STYLE_SPINNER);
        progressDialog.show();
    }

    @Override
    public void hideLoading() {
        if (progressDialog != null) {
            progressDialog.dismiss();
            progressDialog = null;
        }
    }

    @Override
    public void showError(String message) {
        Snackbar.make(rootLayout, message, Snackbar.LENGTH_LONG).show();
    }

    @Override
    public void renderNoteList(List<Note> notes) {
        adapter.setNotes(notes);
    }
}

File: ui/src/main/java/com/example/quarter/ui/MainActivity.kt
package com.example.quarter.ui

import android.app.ProgressDialog
import android.os.Bundle
import com.example.quarter.domain.Note
import com.google.android.material.snackbar.Snackbar
import dagger.android.support.DaggerAppCompatActivity
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.content_main.*
import javax.inject.Inject

class MainActivity : DaggerAppCompatActivity(), MainView {

    @Inject lateinit var presenter: MainPresenter
    @Inject lateinit var adapter: NotesAdapter
    private var progressDialog: ProgressDialog? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setSupportActionBar(toolbar)
        fab.setOnClickListener { presenter.addNote() }
        recyclerView.adapter = adapter
        presenter.setView(this)
        presenter.loadNotes()
    }

    override fun hideLoading() {
        progressDialog?.dismiss()
    }

    override fun showError(message: String) {
        Snackbar.make(rootLayout, message, Snackbar.LENGTH_LONG).show()
    }

    override fun renderNoteList(notes: List<Note>) {
        adapter.notes = notes
    }

    override fun showLoading() {
        progressDialog = ProgressDialog(this).apply {
            setProgressStyle(ProgressDialog.STYLE_SPINNER)
            show()
        }
    }
}
